packages/@aws-cdk/aws-apigateway/lib/integrations/lambda.ts

```diff
@@ -40,14 +54,33 @@ export class LambdaIntegration extends AwsIntegration {
   constructor(handler: lambda.IFunction, options: LambdaIntegrationOptions = { }) {
     const proxy = options.proxy === undefined ? true : options.proxy;
 
+    if (options.asyncProxy && !proxy) {
+      throw new Error('Cannot use `asyncProxy` when `proxy` is set to `false`');
+    }
+
+    let apiHandler = handler;
+    if (proxy && options.asyncProxy) {
+      apiHandler = new lambda.SingletonFunction(cdk.Stack.of(handler), 'AsyncProxy', {
```

Comment:
What happens if there are multiple instances inside the same stack? 

I would extract this into a separate construct and handle singletinity at the stack level there.